week3/homework/src/index.js

```diff
@@ -1,3 +1,192 @@
+app.post('/todos', async (req, res) => {
+  const newTodo = req.body;
+  newTodo.id = uuid();
+  newTodo.done = false;
+  const todos = await readTodos();
+  todos.push(newTodo);
+  await writeTodos(todos);
+  res.json(todos);
+  res.end();
+
+});
+
+//READ ALL TODOs
+app.get('/todos', async (req, res) => {
+  const todos = await readTodos();
+  res.json(todos);
+  res.end();
+});
+
+//READ TODO BY ID
+app.get('/todos/:id', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const todo = todos.find(x => x.id === id);
+    if (typeof todo === 'undefined') {
+      res.json({
+        'Error Message': 'NO such ID !!'
+      });
+    }
+    res.json(todo);
+    res.end();
+
+  } catch (err) {
+    res.json({
+      'Error Name': err.name,
+      'Error Message': err.message
+    });
+  }
+
+});
+
+//DELETE TODO BY ID
+app.delete('/todos/:id', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const index = todos.findIndex(x => x.id === id);
+    if (typeof todos[index] === 'undefined') {
+      res.json({
+        'Error Message': 'NO such ID !!'
+      });
+    } else {
+      todos.splice(index, 1);
+      await writeTodos(todos);
+    }
+    res.json(todos);
+    res.end();
+  } catch (err) {
+    res.json({
+      'Error Name': err.name,
+      'Error Message': err.message
+    });
+  }
+
+});
+
+//DELETE ALL TODOs
+app.delete('/todos', async (req, res) => {
+  await writeTodos([]);
+  res.json([]);
+  res.end();
+});
+
+//UPDATE TODO BY ID
+app.put('/todos/:id', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const newTodo = req.body;
+    newTodo.id = id;
+    const index = todos.findIndex(x => x.id === id);
+    if (typeof todos[index] === 'undefined') {
+      res.json({
+        'Error Message': 'NO such ID !!'
+      });
+    } else {
+      newTodo.done = todos[index].done;
+      todos.splice(index, 1, newTodo);
+      await writeTodos(todos);
+    }
+    res.json(todos);
+    res.end();
+  } catch (err) {
+    res.json({
+      'Error Name': err.name,
+      'Error Message': err.message
+    });
+  }
+
+});
+
+//markAsDone TODO
+app.post('/todos/:id/done', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const newTodos = todos.find(todos => todos.id === id);
+    if (typeof newTodos === 'undefined') {
+      res.json({
+        'Error Message': 'NO such ID !!'
+      });
+    } else {
+      newTodos.done = true;
+      await writeTodos(todos);
+    }
+    res.json(todos);
+    res.end();
+  } catch (err) {
+    res.json({
+      'Error Name': err.name,
+      'Error Message': err.message
+    });
+  }
+
+});
+
+//markAsNotDone TODO
+app.delete('/todos/:id/done', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const newTodos = todos.find(todos => todos.id === id);
```

Comment:
Besides the changes you need to make, I see the same lines of code repeating. Maybe you could have some functions to help you do these actions on the todos. For instance a `findTodo` function which would accept the `request` as a param and return either throw an error (which you could `catch` 😉 ) or the `todo` that was found.